mtbl/block.c

```diff
@@ -291,24 +295,43 @@ block_iter_seek(struct block_iter *bi, const uint8_t *target, size_t target_len)
 		}
 	}
 
-	/* binary search */
-	while (left < right) {
-		uint32_t mid = (left + right + 1) / 2;
-		if (compare_restart_point(bi, mid, target, target_len) < 0) {
-			/* key at "mid" is smaller than "target", therefore all
-			 * keys before "mid" are uninteresting
-			 */
-			left = mid;
-		} else {
-			/* key at "mid" is larger than "target", therefore all
-			 * keys at or after "mid" are uninteresting
-			 */
-			right = mid - 1;
+	/* binary search, if not already located. */
+	if (left + 1 < right) {
+		while (left < right) {
+			uint32_t mid = (left + right + 1) / 2;
+			if (compare_restart_point(bi, mid, target, target_len) < 0) {
+				/* key at "mid" is smaller than "target", therefore all
+				* keys before "mid" are uninteresting
+				*/
+				left = mid;
+			} else {
+				/* key at "mid" is larger than "target", therefore all
+				* keys at or after "mid" are uninteresting
+				*/
+				right = mid - 1;
+			}
 		}
 	}
 
+	/* Desired entry is in the same restart-block as "current" key. */
+	if (start_ri == left) {
+		/* Check current entry against the target. */
+		int cmp = bytes_compare(ubuf_data(bi->key), ubuf_size(bi->key), target, target_len);
+
+		if (cmp == 0)
+			return;
```

Comment:
Can calculating `cmp` and returning can be moved to the very top of the function for additional optimization points? Assuming seeking to where it already is actually happens all the time in real life production.

Reply:
For a data block, the `bi->key` key will be the same key returned by the preceding `mtbl_iter_next()`, and is unlikely to be the target of an `mtbl_iter_seek()` operation.

For an index block the key will be a minimal length key strictly between two successive mtbl keys, and is computed from those two specific keys. Thus, it is also unlikely to be the target of `mtbl_iter_seek`.